### PR TITLE
[AutoDiff] Add #chainableGradient differential operator for seedable gradient

### PR DESCRIPTION
**Pull request by rxwei**
`#chainableGradient` takes a function and returns a function, which takes original parameters and a backpropagated adjoint, and computes vector-Jacobian products.

```swift
func f(_ x: Float) -> Float { ... }
let y = f(x)
let dz_dy = ...
let dz_dx = #chainableGradient(f)(x, dz_dy)
```

`#chainableGradient` corresponds to the `gradient [seedable]` instruction in SIL.

Sema and libSyntax are to be done separately. This is a random patch to unblock the progress towards a way to express seedable derivatives in Swift source code.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Review by dan-zheng — APPROVED** [review]
Sweet!

**Comment by dan-zheng** [comment]
I'm happy to tackle Sema and Syntax for `#chainableGradient` if you'd like to work on other parts of autodiff. Let me know!

**Comment by rxwei** [comment]
@dan-zheng Thx. Don't worry about it today. Let's try get more deabstraction bugs fixed.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow
